### PR TITLE
impl(oauth2): add anonymous and access token credentials

### PR DESCRIPTION
**Pull request by scotthart**
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8250)

### CODE REVIEW

**Review by coryan — APPROVED** [review]
nice code coverage

> Reviewed 9 of 9 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8250#-:-Mv6G9c8C7BhtfT04ZRl:b1nw0e5)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @scotthart)

File: google/cloud/internal/oauth2_anonymous_credentials_test.cc [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+#include "google/cloud/testing_util/status_matchers.h"
+#include <gmock/gmock.h>
+
+namespace google {
+namespace cloud {
+namespace oauth2_internal {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+namespace {
+
+class AnonymousCredentialsTest : public ::testing::Test {};
```
Comment (devbww) [inline google/cloud/internal/oauth2_anonymous_credentials_test.cc:25]: Does this buy anything over just using `TEST()`?

**Review by devbww — COMMENTED** [review]
You may care to do some IWYU checking for things like pair/make_pair, chrono, StatusOr.

**Review by scotthart — COMMENTED** [review]
Added appropriate headers to the .h files.

> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8250#-:-MvBHeovBFtwAeJPSJRE:ba3zy8v)* status: 4 of 9 files reviewed, 1 unresolved discussion (waiting on @coryan and @devbww)

___
*[google/cloud/internal/oauth2_anonymous_credentials_test.cc, line 25 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8250#-Mv5yf30JAdV26HY3pBM-r1-25:-MvBHr5S1Oe1cZQTQNqT:b-grh9qx) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/683178e6bb3bdeba83ff3a0d8901642e3c527a95/google/cloud/internal/oauth2_anonymous_credentials_test.cc#L25)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

Does this buy anything over just using `TEST()`?
</blockquote></details>

Changed to TEST()

**Pull request merged**